### PR TITLE
adding support for libc's "fgets" and "fgetc" needed for Ada language

### PR DESCRIPTION
**Pull request by kg8280**
These are libc functions used from Ada runtime libraries to read from standard input.

### CODE REVIEW

File: runtime/LibcWrappers.cpp [hunk 1]
```diff
@@ -246,6 +246,26 @@ size_t SYM(fread)(void *ptr, size_t size, size_t nmemb, FILE *stream) {
...
+char *SYM(fgets)(char *str, int n, FILE *stream) {
+  tryAlternative(str, _sym_get_parameter_expression(0), SYM(fgets));
+  tryAlternative(n, _sym_get_parameter_expression(1), SYM(fgets));
+
+  auto result = fgets(str, n, stream);
+  _sym_set_return_expression(nullptr);
+
+  if (fileno(stream) == inputFileDescriptor) {
+    // Reading symbolic input.
+    ReadWriteShadow shadow(str, sizeof(char) * n);
```
Comment (sebastianpoeplau) [inline runtime/LibcWrappers.cpp:258]: According to the man page, `fgets` reads at most `n-1` characters, and it also stops at EOF or after a newline. I believe that using `n` here unconditionally may lead to a wrong increment of `inputOffset` below (e.g., if `fgets` is called with a large buffer but ends up reading just a couple of characters before it encounters the end of the line). Shouldn't we use `strlen(str)` instead? The same concern applies to the case below where we erase the shadow.

File: runtime/LibcWrappers.cpp [hunk 2]
```diff
@@ -246,6 +246,26 @@ size_t SYM(fread)(void *ptr, size_t size, size_t nmemb, FILE *stream) {
   return result;
 }
 
+char *SYM(fgets)(char *str, int n, FILE *stream) {
+  tryAlternative(str, _sym_get_parameter_expression(0), SYM(fgets));
+  tryAlternative(n, _sym_get_parameter_expression(1), SYM(fgets));
+
+  auto result = fgets(str, n, stream);
+  _sym_set_return_expression(nullptr);
```
Comment (sebastianpoeplau) [inline runtime/LibcWrappers.cpp:254]: `fgets` returns `str` on success, so I suppose we should do `_sym_set_return_expression(_sym_get_parameter_expression(0))` in that case to make sure that the analysis understands the return value if it's used later on.

**Review by sebastianpoeplau — COMMENTED** [review]
Great, thanks for the pull request! I've added a couple of comments; if you don't have any objections, I can address them and merge the changes.

**Comment by kg8280** [comment]
Hi Sebastian,

Thank you for your review and for catching up on the mistakes. I am still in the process of understanding how things work :).
I made all the changes. Please, let me know if now everything is OK.

Thank you,
Kyriakos

File: runtime/LibcWrappers.cpp (hunk 1 above)
Comment (kg8280) [reply runtime/LibcWrappers.cpp]: I applied the suggested changes :)

File: runtime/LibcWrappers.cpp (hunk 2 above)
Comment (kg8280) [reply runtime/LibcWrappers.cpp]: I applied the suggested changes :)

**Comment by sebastianpoeplau** [comment]
Sorry for the long delay! All is good - I've rebased and merged, thanks a lot for the contribution!

**Pull request merged**